### PR TITLE
feat(manager): support getting and putting files stored in S3

### PR DESCRIPTION
**Pull request by graczhual**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2p29v2p 二进制文件读取对接: S3](https://app.clickup.com/t/2p29v2p)

File: graviti/manager/policy.py [hunk 1]
```diff
@@ -293,3 +293,169 @@ def put_object(self, key: str, path: Path, _allow_retry: bool = True) -> None:
...
                 self.put_object(key, path, False)
             raise error from None
+
+
+class S3ObjectPolicyManager(ObjectPolicyManager):
+    """The basic structure of the object policy of the dataset stored in S3."""
+
+    _RETRY_CODE = {"InvalidAccessKeyId", "AccessDenied"}
+    _GET_HASHED_PAYLOAD = sha256(b"").hexdigest()
+    _PUT_HASHED_PAYLOAD = "UNSIGNED-PAYLOAD"
```
Comment (linjiX) [inline graviti/manager/policy.py:303]: the hashed payload is bound with verb, maybe we can use a dict to store them?

File: graviti/manager/__init__.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
     AZUREObjectPolicyManager,
     ObjectPolicyManager,
     OSSObjectPolicyManager,
+    S3ObjectPolicyManager,
```
Comment (linjiX) [inline graviti/manager/__init__.py:15]: will it be better to set the commit scope to `manager`

File: graviti/manager/policy.py [hunk 3]
```diff
@@ -293,3 +293,173 @@ def put_object(self, key: str, path: Path, _allow_retry: bool = True) -> None:
...
                 self.put_object(key, path, False)
             raise error from None
+
+
+class S3ObjectPolicyManager(ObjectPolicyManager):
+    """The basic structure of the object policy of the dataset stored in S3."""
+
+    _RETRY_CODE = {"InvalidAccessKeyId", "AccessDenied"}
+    _HASHED_PAYLOAD = {
+        "GET": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
```
Comment (linjiX) [inline graviti/manager/policy.py:303]: this string is meaningless for human, maybe the old presentation is better.

File: graviti/manager/policy.py [hunk 4]
```diff
@@ -293,3 +293,173 @@ def put_object(self, key: str, path: Path, _allow_retry: bool = True) -> None:
...
+            self._put_policy = put_policy
+        return self._put_policy
+
+    @staticmethod
+    def _get_canonical_request(
+        policy: Dict[str, Any], verb: str, key: str, x_amz_date: str, hashed_payload: str
+    ) -> str:
+        canonical_headers = (
+            f"host:{policy['bucket']}.{policy['endpoint']}\n"
+            f"x-amz-content-sha256:{hashed_payload}\n"
```
Comment (linjiX) [inline graviti/manager/policy.py:345]: shall we also get the `hashed_payload` from the `dict` here?

**Review by linjiX — CHANGES_REQUESTED** [review]

**Review by linjiX — APPROVED** [review]

**Comment by GravitiRob** [comment]
Task linked: [CU-2p29v0w 二进制文件上传对接: S3](https://app.clickup.com/t/2p29v0w)

**Pull request merged**